[FLINK-10282][rest] Separate REST and Dispatcher RPC thread pools

zentol opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This PR separates the REST and DispatcherRPC thread pools. Basically the `WebMonitorEndpoint` creates it's own `ExecutorService` based on the `rest.server.numThreads` option, instead of being passed the Dispatcher RPC's `Executor`.

## Brief change log

* add `RestOptions#SERVER_NUM_THREADS`
*  extend `RestServerEndpointConfiguration` to process new option (this change is questionable since the `RestServerEndpoint` never uses this option)
* create `ExecutorService` in `WebMonitorEndpoint`.

## Verifying this change

I've added a basic test to cover the behavior of the `RestServerEndpointConfiguration`.

Testing the executor service itself is possible (submit N blocking tasks, count how many end up blocking, assert numBlocking==configuredNumThreads) but getting it stable is tricky.

## Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? (docs)

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestServerEndpointConfigurationTest.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Tests for the {@link RestServerEndpointConfiguration}.
+ */
+public class RestServerEndpointConfigurationTest {
```
`extends TestLogger` missing

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -714,6 +717,7 @@ public void startInternal() throws Exception {
 	@Override
 	protected CompletableFuture<Void> shutDownInternal() {
 		executionGraphCache.close();
+		ExecutorUtils.gracefulShutdown(10, TimeUnit.SECONDS, executor);
```
I think we should shut down the `executor` after shutting down the handlers. Otherwise we might run into the situation that a handler tries to add something to a closed `Executor`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -180,7 +183,7 @@ public WebMonitorEndpoint(
 		this.resourceManagerRetriever = Preconditions.checkNotNull(resourceManagerRetriever);
 		this.transientBlobService = Preconditions.checkNotNull(transientBlobService);
-		this.executor = Preconditions.checkNotNull(executor);
+		this.executor = Executors.newFixedThreadPool(endpointConfiguration.getNumThreads(), new ExecutorThreadFactory("Flink-" + getClass().getSimpleName()));
```
Could we still keep it like before that we pass in an `executor`? This makes testing much easier because you have control over the actual `ExecutorService` which you can pass in. What we would need is a kind of factory which is used by the `ClusterEntrypoint` to create the `ExecutorService` and the `WebMonitorEndpoint`. What do you think?

zentol replied:
Who would then be responsible for shutting it down?

zentol replied:
Given that no test was making use of that feature I'd be inclined to keep the PR as is. If all our tests so far could be implemented without using this feature, what test could possibly require it?

tillrohrmann replied:
It would be the `WebMonitorEndpoint` task to do so. The contract would be that the `WebMonitorEndpoint` is responsible for managing this executor.

In general, it is a good idea to not create things in the constructor of a component, because it makes it really hard to reuse it in a different environment. Instead, it is imo better to pass in instances of some interface. That way, we don't couple components.

Even if we don't do it right now, we should not remove functionality which allowed it in the first place.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:
```diff
@@ -180,7 +183,7 @@ public WebMonitorEndpoint(
 		this.resourceManagerRetriever = Preconditions.checkNotNull(resourceManagerRetriever);
 		this.transientBlobService = Preconditions.checkNotNull(transientBlobService);
-		this.executor = Preconditions.checkNotNull(executor);
+		this.executor = Executors.newFixedThreadPool(endpointConfiguration.getNumThreads(), new ExecutorThreadFactory("Flink-" + getClass().getSimpleName()));
```
This could be a follow up: Can we start the threads with `Thread.MIN_PRIORITY` in order to avoid congestion wrt the rpc threads?

zentol replied:
Shouldn't we be worried about the other end of worst-case scenarios: the REST API (and thus client and WebUI) becoming completely unresponsive?

zentol replied:
A simpler alternative might be to bump the Dispatcher priority instead; then we don't have to adjust every other component (like metrics).

tillrohrmann replied:
In general, I think the REST API is less important than the cluster components. Bumping the priority of the `Dispatcher` could also be a solution.

tillrohrmann commented:
Thanks for addressing some of my components @zentol. I think we should still change how the `ExecutorService` is instantiated. Instead of doing it in the constructor it should happen imo outside of the component. That way, we don't couple different components (e.g. which `ExecutorService` is used by the `WebMonitorEndpoint`). Moreover, it was already like this before and, thus, we would introduce a regression (in terms of architecture).

yanghua commented:
hi @zentol, It seems like this PR should be linked to [FLINK-10282](https://issues.apache.org/jira/browse/FLINK-10282)?

zentol commented:
@tillrohrmann Could you take another look?

## Brief change log

* reverted removal of Executor constructor argument, but changed type to ExecutorService
* reverted chances to `RestServerEndpointConfiguration`
* introduced a `Builder` for `ExecutorThreadFactory`
* made thread-priority configurable in `ExecutorThreadFactory`
* added ´WebMonitorEndpoint.createExecutorService` as a default factory

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -341,7 +343,9 @@ public void start() throws Exception {
-					commonRpcService.getExecutor(),
+					WebMonitorEndpoint.createExecutorService(
+						configuration.getInteger(RestOptions.SERVER_NUM_THREADS),
+						"DispatcherRestEndpoint"),
```
In the `MiniCluster` case, it might be justifiable to use the `commonRpcService.getExecutor`. In particular, if it is configured to use a shared rpc service/`useSingleRpcService == true`. For this work, we would need a `CloseIgnoringExecutorService` which wraps an `Executor` and ignores the shutdown call. This can also be a follow up.

zentol replied:
Would the following be all that we need?
```
	private static class CloseIgnoringExecutorService extends AbstractExecutorService {

		private final Executor executor;

		public CloseIgnoringExecutorService(Executor executor) {
			this.executor = executor;
		}

		@Override
		public void shutdown() {
		}

		@Override
		public List<Runnable> shutdownNow() {
			return Collections.emptyList();
		}

		@Override
		public boolean isShutdown() {
			return false;
		}

		@Override
		public boolean isTerminated() {
			return false;
		}

		@Override
		public boolean awaitTermination(long timeout, TimeUnit unit) {
			return false;
		}

		@Override
		public void execute(Runnable command) {
			executor.execute(command);
		}
	}
```

zentol replied:
Is this about making the MIniCluster more light-weight? If so, couldn't we also reduce the thread-pool size to 1 instead?
Re-using another executor means that it will both differ in behavior and thread-debugging (different names) compared to production.

tillrohrmann replied:
Almost. I think we should have a field `terminationFuture`:
```
private static class CloseIgnoringExecutorService extends AbstractExecutorService {

	private final Executor executor;

	private final CompletableFuture<Void> terminationFuture = new CompletableFuture<>();

	public CloseIgnoringExecutorService(Executor executor) {
		this.executor = executor;
	}

	@Override
	public void shutdown() {
		terminationFuture.complete(null);
	}

	@Override
	public List<Runnable> shutdownNow() {
		shutdown();
		return Collections.emptyList();
	}

	@Override
	public boolean isShutdown() {
		return terminationFuture.isDone();
	}

	@Override
	public boolean isTerminated() {
		return terminationFuture.isDone();
	}

	@Override
	public boolean awaitTermination(long timeout, TimeUnit unit) {
		try {
			terminationFuture.get(timeout, unit);
			return true;
		} catch (TimeoutException e) {
			return false;
		}
	}

	@Override
	public void execute(Runnable command) {
		if (terminationFuture.isDone()) {
			throw new RejectedExecutionException();
		} else {
			executor.execute(command);
		}
	}
}
```

tillrohrmann replied:
Yes that is the intention. Decreasing the size of the thread pool would also be a good option.
